Packaging DeviceConfiguration and SimpleConfig Applications

kb1lqc opened the pull request:
This PR updates the Device Configuration and SimpleConfig applications to be imported into the python package for `faraday`. It may not be implemented the cleanest (see --fxxxxx options) or the fact that simpleconfig isn't integrated with deviceconfiguration but it is pretty straightforward and we can improve in later updates. The distribution files have been uploaded to TestPyPi to allow pip installation from the test repository. Alternatively you can install this git branch in editable mode with `pip install -e .` from the main folder.

@kb1lqd @el-iso  @jbemenheiser  @reillyeon

# Change Log
- Moved deviceconfiguration into `--faraday-deviceconfiguration`
- Moved simpleconfig into `--faraday-simpleconfig`
- Moved configuration files `deviceconfiguration.sample.ini` and `faraday_config.sample.ini` into etc/faraday
- Updated setup.cfg to reflect new version + new etc/files with commands for these two programs
- Changed `deviceconfiguration.ini` [DEVICES] options for use with one radio at a time. This simplified the code and really we only want to program one radio at a time right now. Future updates can expand this.
- Added [PROXY] section to `deviceconfiguration.sample.ini`
- Added appropriate loggers for Device Configuration and SimpleConfig programs to `loggingConfig.ini`
- Updated **ALL programs** loggers in `loggingConfig.ini` to INFO level
- Added command line options for `faraday-deviceconfiguration` and `faraday-simpleconfig`
- Added functionality to simpleconfig to read only and not program

# Command Line Options
## Device Configuration
```
$ faraday-deviceconfiguration --help
usage: faraday-deviceconfiguration [-h] [--init-config]
                                   [--init-faraday-config] [--start]
                                   [--proxycallsign PROXYCALLSIGN]
                                   [--proxynodeid PROXYNODEID]
                                   [--faradayconfig] [--callsign CALLSIGN]
                                   [--nodeid NODEID] [--redledtxon]
                                   [--redledtxoff] [--unitconfigured]
                                   [--unitunconfigured] [--gpiop3on GPIOP3ON]
                                   [--gpiop3off GPIOP3OFF] [--gpiop3clear]
                                   [--gpiop4on GPIOP4ON]
                                   [--gpiop4off GPIOP4OFF] [--gpiop4clear]
                                   [--gpiop5on GPIOP5ON]
                                   [--gpiop5off GPIOP5OFF] [--gpiop5clear]
                                   [--gpiop5 GPIOP5]
                                   [--bootfrequency BOOTFREQUENCY]
                                   [--bootrfpower BOOTRFPOWER]
                                   [--latitude LATITUDE]
                                   [--longitude LONGITUDE]
                                   [--latitudedir LATITUDEDIR]
                                   [--longitudedir LONGITUDEDIR]
                                   [--altitude ALTITUDE] [--gpsbooton]
                                   [--gpsbootoff] [--gpsenabled]
                                   [--gpsdisabled] [--uarttelemetryenabled]
                                   [--uarttelemetrydisabled]
                                   [--rftelemetryenabled]
                                   [--rftelemetrydisabled]
                                   [--uartinterval UARTINTERVAL]
                                   [--rfinterval RFINTERVAL]

Device Configuration application provides a Flask server to program Faraday
radios via an API

optional arguments:
  -h, --help            show this help message and exit
  --init-config         Initialize Device Configuration configuration file
  --init-faraday-config
                        Initialize Faraday configuration file
  --start               Start device configuration server
  --proxycallsign PROXYCALLSIGN
                        Set Proxy Faraday callsign to connect to and program
  --proxynodeid PROXYNODEID
                        Set Proxy Faraday nodeid to connect to and program
  --faradayconfig       Display Faraday configuration file contents
  --callsign CALLSIGN   Set Faraday radio callsign
  --nodeid NODEID       Set Faraday radio nodeid
  --redledtxon          Set Faraday radio RED LED during RF transmissions ON
  --redledtxoff         Set Faraday radio RED LED during RF transmissions OFF
  --unitconfigured      Set Faraday radio configured bit ON
  --unitunconfigured    Set Faraday radio configured bit OFF
  --gpiop3on GPIOP3ON   Set Faraday radio GPIO port 3 bits on, specify bit to
                        turn ON
  --gpiop3off GPIOP3OFF
                        Set Faraday radio GPIO port 3 bits on, specify bit to
                        turn OFF
  --gpiop3clear         Reset Faraday radio GPIO port 3 bits to OFF
  --gpiop4on GPIOP4ON   Set Faraday radio GPIO port 4 bits on, specify bit to
                        turn ON
  --gpiop4off GPIOP4OFF
                        Set Faraday radio GPIO port 4 bits on, specify bit to
                        turn OFF
  --gpiop4clear         Reset Faraday radio GPIO port 4 bits to OFF
  --gpiop5on GPIOP5ON   Set Faraday radio GPIO port 5 bits on, specify bit to
                        turn ON
  --gpiop5off GPIOP5OFF
                        Set Faraday radio GPIO port 5 bits on, specify bit to
                        turn OFF
  --gpiop5clear         Reset Faraday radio GPIO port 5 bits to OFF
  --gpiop5 GPIOP5       Set Faraday radio fgpio_p5
  --bootfrequency BOOTFREQUENCY
                        Set Faraday radio boot frequency
  --bootrfpower BOOTRFPOWER
                        Set Faraday radio boot RF power
  --latitude LATITUDE   Set Faraday radio default latitude. Format "ddmm.mmmm"
  --longitude LONGITUDE
                        Set Faraday radio default longitude. Format
                        "dddmm.mmmm"
  --latitudedir LATITUDEDIR
                        Set Faraday radio default latitude direction (N/S)
  --longitudedir LONGITUDEDIR
                        Set Faraday radio default longitude direction (E/W)
  --altitude ALTITUDE   Set Faraday radio default altitude in meters. Maximum
                        of 17999.99 Meters
  --gpsbooton           Set Faraday radio GPS boot power ON
  --gpsbootoff          Set Faraday radio GPS boot power OFF
  --gpsenabled          Set Faraday radio GPS use ON
  --gpsdisabled         Set Faraday radio GPS use OFF
  --uarttelemetryenabled
                        Set Faraday radio UART Telemetry ON
  --uarttelemetrydisabled
                        Set Faraday radio UART Telemetry OFF
  --rftelemetryenabled  Set Faraday radio RF Telemetry ON
  --rftelemetrydisabled
                        Set Faraday radio RF Telemetry OFF
  --uartinterval UARTINTERVAL
                        Set Faraday radio UART telemetry interval in seconds
  --rfinterval RFINTERVAL
                        Set Faraday radio RF telemetry interval in seconds

```

## SimpleConfig
```
$ faraday-simpleconfig --help
usage: faraday-simpleconfig [-h] [--read]

SimpleConfig sends a request to faraday-deviceconfiguration to initiate a POST
or GET command resulting in programming a Faraday radio and/or reading its
FLASH memory configuration

optional arguments:
  -h, --help  show this help message and exit
  --read      Read FLASH configuration only, do not program
```

# Installing from TestPyPi
As of this PR creation the latest package on TestPyPi is [faraday-0.0.1014.dev1068](https://testpypi.python.org/pypi/faraday/0.0.1014.dev1068)

Install with `pip install --no-cache-dir --extra-index-url https://testpypi.python.org/pypi faraday`

Ensure you are installing this correct version per this PR

kb1lqc committed:
Initial commit of deviceconfiguration update

Moving the files around to support importing and wrapper .py scripts just
like other applications. Eventually moving towards Python Packaging and
command line functionality.

kb1lqc committed:
Initial Working File Structure

Wrapper script in /Applications/DeviceConfiguration successfully launches
the /faraday/deviceconfiguration.py. Also, loggingconfig.ini is imported
correctly with new method in packaged format.

kb1lqc committed:
Setup.cfg recognizes device configuration

Setup.cfg imports device configuration functionality and the revision of
code was updated to 0.0.1014

kb1lqc committed:
Moved sample Device Configuration INI

The sample configuration file was moved into etc/faraday and a test
configuration file was successfully imported with this code from the
/etc/faraday folder.

kb1lqc committed:
Removed old configuration import code

Commented out code deleted.

kb1lqc committed:
Successfully programmed Faraday

Using a manually run simpleconfig.py and manually generated INI files this
programmed a radio.

kb1lqc committed:
Initial command line arguments added

Initialization of configuration file works and basic structure of command
line arguments added.

kb1lqc committed:
Updated To Program configuration file

deviceconfiguration.ini now queried for callsign and node ID by simpleui
in effort to ease move over to self-contained code in device
configuration. Also, callsign and nodeid command line options now work.

kb1lqc committed:
Updated --program in command linbe

Importing simpleui functionality and mostly there. I had to add a few
configuration options to the INI file for simplicity and consolidation.
However, this does not work. I'm still getting HTTP 400 response when I
try to program. Probably due to how device configuration is responding to
proxy. Something funky is up.

kb1lqc committed:
Removed --program functionality

It's not terribly useful since it only acts before the Flask server starts
up for device configuration. I could implement some sort of state machine
to do this but at this point it should probably be considered moving this
stuff into Proxy anyways. Therefore I'm going to implement the
simpleconfig + deviceconfig workflow for now.

kb1lqc committed:
Implemented setup.cfg simpleconfig

Moved files for simpleconfig around and created a /faraday/simpleconfig.py
implementation. Also added the function to setup.cfg but will need to move
/etc/faraday files next as well.

kb1lqc committed:
Added init-faraday-config

Copy over faraday_config.sample.ini to faraday_config.ini using commands
in faraday-deviceconfig.

kb1lqc committed:
Updated simpleconfig for correct imports

Added path importing functions to import logging and deviceconfig INI
files correctly.

kb1lqc committed:
Basic simpleconfig functionality

Working to program the radio however I need to finish cleaning up the
document for print statements and overall operation. Also, I do not
currently allow command line editing of the configuration files so that
needs to happen too.

kb1lqc committed:
Updated prints to logging statements for read

Simpleconfig read after programming to see actual Faraday firmware
variables now uses logging statements and correct string formatting.

kb1lqc committed:
Added configuration file check

Check for configuration file and if not found nicely ask user to
initialize.

kb1lqc committed:
Added Faraday configuration options

Added argparse command line option for faraday_config.ini options.
Performs some basic checking like int/float and defaults to the current
settings in the faraday_config.sample.ini such as configbootbitmask and
gps_boot_bit are true unless turned off by command.

This does not check for properly formatted/length GPS lat/lon values or
other fields. Will have to assess that in a future commit or issue.

kb1lqc committed:
Fixed up simpleconfig

Removed logging statements that weren't needed and added a few more info
logging statements.

kb1lqc committed:
Added callsign + nodeid to simpleonfig output

Wasn't printing the response of simpleconfig query for programmed faraday
callsign and nodeid. Now I am.

kb1lqc committed:
Cleaned up random comment

Just some cleaning, nothing to see here.

kb1lqc committed:
Added helpful tips to configuration options

Added format and maximum values to Latitude/Longitude as well as Altitude
in the configuration help options.

kb1lqc committed:
Updated simpleconfig and deviceconfig for Pytest

Tested with pytest for PEP8 warnings we currently check for and fixed all
indicated issues.

kb1lqc committed:
Updated help functions

Improved text indicating whether a setting will turn a function on or off
discretely.

kb1lqc committed:
Updated simpleconfig logger use

SimpleConfig incorrectly used logging instead of logger for its info
and error statements. This was changed. I also updated the
deviceconfiguration logger name.

kb1lqc committed:
Updated all loggingconfig.ini to INFO level

This is actually for ALL programs such as Proxy, Telemetry, APRS,
SimpleUI, DeviceConfiguration, and SimpleUI. I changed the logging level
to INFO. These may not be used in all cases (defaults to root setting
instead of specified program level) but that's an issue ticket in the
future as we identify them.

kb1lqc committed:
Added SimpleConfig --read option

Let's you just read the configuration instead of programming the Faraday
radio

kb1lqc committed:
Removed read URL query hard-coding

Updated the request.get() command for reading the FLASH configuration to
use the configured PROXY settings in the INI file instead of hard-coding
the URL and port.

reillyeon commented:
Skimming through the help info it seems unnecessary to prefix all the options with "f".

kb1lqc commented:
@reillyeon I actually prefix the options which change actual faraday programmed memory with an `--f`. Therefore as it currently stands `--callsign` changes the device configuration program callsign setting while `--fcallsign` changes the intended callsign to be programmed to faraday. Any other suggestions?

kb1lqd commented:
@kb1lqc Why not just use the terms `--proxycallsign` to change local connection callsign and just `--callsign` for the device configuration update callsign? I agree with @reillyeon that prepending `f` before each command is unnecessary.

kb1lqc commented:
OK I'll update shortly @kb1lqd  @reillyeon

kb1lqd commented:
@kb1lqc How does `faraday-deviceconfiguration --fconfigboot` work? I can't seem to get it to do anything. Also this should not simply be the only way to edit the configbootbitmask. We should have a way to set the bitmask manually to a value (byte) and/or also support the "RF RED LED" bit 1 as well.

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,440 @@
+parser = argparse.ArgumentParser(description='Device Configuration application provides a Flask server to program Faraday radios via an API')
+parser.add_argument('--init-config', dest='init', action='store_true', help='Initialize Device Configuration configuration file')
+parser.add_argument('--init-faraday-config', dest='initfaraday', action='store_true', help='Initialize Faraday configuration file')
+parser.add_argument('--callsign', help='Set Proxy Faraday callsign to connect to and program')
```
This should be `--proxycallsign` or similar and allow the removal of `f` prepending the other arguments.

kb1lqc replied:
Changed in https://github.com/FaradayRF/Faraday-Software/pull/219/commits/1556fe743c340aa67beec7e3e4849fb19021df27

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,440 @@
+# Faraday Configuration
+parser.add_argument('--fcallsign', help='Set Faraday radio callsign')
+parser.add_argument('--fnodeid', type=int, help='Set Faraday radio nodeid')
+parser.add_argument('--fconfigboot', action='store_false', help='Set Faraday radio config boot bit OFF')
```
1) This doesn't seem to work... Also how do you toggle this back ON? (This is BIT 0)
2) Need to add for BIT 1 the toggling of the RED LED action on RF TX

If both bits are high then config boot bitmask = 3 (00000011)

kb1lqc replied:
@kb1lqd 

1. It defaults to 1 and if you specify `--configboot` the INI file is changed to a zero.
2. Looks like I implemented it wrong haha....

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,440 @@
+parser.add_argument('--flongitudedir', help='Set Faraday radio default longitude direction (E/W)')
+parser.add_argument('--faltitude', type=float, help='Set Faraday radio default altitude in meters. Maximum of 17999.99 Meters')
+# Purposely do not allow editing of GPS altitude units
+parser.add_argument('--fgpsboot', action='store_false', help='Set Faraday radio GPS boot power OFF')
```
How do you toggle back on? I suggest either allowing argument options `0` and '1' or creating an `ON` and `OFF` function arguments.

kb1lqc replied:
Defaults to ON. I should update the help description to state this.

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,440 @@
+parser.add_argument('--faltitude', type=float, help='Set Faraday radio default altitude in meters. Maximum of 17999.99 Meters')
+# Purposely do not allow editing of GPS altitude units
+parser.add_argument('--fgpsboot', action='store_false', help='Set Faraday radio GPS boot power OFF')
+parser.add_argument('--fgps', action='store_true', help='Set Faraday radio GPS use ON')
```
How do you toggle back on? I suggest either allowing argument options `0` and '1' or creating an `ON` and `OFF` function arguments.

kb1lqc replied:
Defaults to OFF, I should add more to the help statement.

kb1lqc committed:
Updated deviceconfiguration options to remove f

the leading --f in options to program faraday were removed and
callsign-nodeid for deviceconfiguration configuration was prepended with
--proxy.

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,440 @@
+# Purposely do not allow editing of GPS altitude units
+parser.add_argument('--fgpsboot', action='store_false', help='Set Faraday radio GPS boot power OFF')
+parser.add_argument('--fgps', action='store_true', help='Set Faraday radio GPS use ON')
+parser.add_argument('--fuarttelemetry', action='store_false', help='Set Faraday radio UART Telemetry OFF')
```
How do you toggle back on? I suggest either allowing argument options 0 and '1' or creating an ON and OFF function arguments.

kb1lqc replied:
Defaults to ON, will add statement to --help description

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,440 @@
+parser.add_argument('--fgpsboot', action='store_false', help='Set Faraday radio GPS boot power OFF')
+parser.add_argument('--fgps', action='store_true', help='Set Faraday radio GPS use ON')
+parser.add_argument('--fuarttelemetry', action='store_false', help='Set Faraday radio UART Telemetry OFF')
+parser.add_argument('--frftelemetry', action='store_true', help='Set Faraday radio RF Telemetry ON')
```
How do you toggle back off? I suggest either allowing argument options 0 and '1' or creating an ON and OFF function arguments.

kb1lqc replied:
Defaults to OFF, will add help information about this.

kb1lqc commented:
@kb1lqd maybe I misunderstood the value. You can just turn it ON/OFF. Not placing `--configboot` in the command will default to setting the INI file to a value of 1 and by specifying `--configboot` you change it to a zero. This is explained in the `--help` option.

kb1lqc commented:
@kb1lqd just responded to your questions. Overall I basically made any ON/OFF option in _faraday.ini_ default to what we have in the _faraday.sample.ini_ configuration.

I do think I completely miss-implemented the `--configboot` option. Should we converse more about how to implement each bit now or in a later PR and remove the option for the time being?

kb1lqd commented:
@kb1lqc Yeah this is not a fun way to do it for several reason:
* Editing multiple units with multiple settings (or making a mistake and wanting a different option) means re initializing the whole file.
* Defaulting to a standard setting is fine but any bit toggling NEED's both ON and OFF due to the above

Also, the bitmask bytes are literally bitmask bytes so all bits in use need to be editable. This is not acceptable for a later PR since things like the RED LED toggle are very useful and should be supported off the bat as they currently are.

kb1lqc committed:
Updated ON/OFF functionality of options

All options that are ON/OFF are now explicitely so. Values in the
configuration file will not be touched if not specified.

kb1lqc commented:
@kb1lqd what are the bitmask values? what does each bit represent?

kb1lqc committed:
Added --start option

Check for --start option and if not found the program is exited after
configuration updates have been made.

kb1lqd commented:
[Flash_Info_D_Memory_Allocations.xlsx](https://github.com/FaradayRF/Faraday-Software/files/1118082/Flash_Info_D_Memory_Allocations.xlsx)


We should bring a lot of this non-public facing documentation to public as it should. I think this fell through the cracks until we started needing to reference it again...

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,450 @@
+parser.add_argument('--bootrfpower', type=int, help='Set Faraday radio boot RF power')
+parser.add_argument('--latitude', type=float, help='Set Faraday radio default latitude. Format \"ddmm.mmmm\"')
+parser.add_argument('--longitude', type=float, help='Set Faraday radio default longitude. Format \"dddmm.mmmm\"')
+parser.add_argument('--latitudedir', help='Set Faraday radio default latitude direction (N/S)')
```
This allows other characters besides N/S. Filter?

kb1lqd commented on faraday/deviceconfiguration.py:
```diff
@@ -0,0 +1,450 @@
+parser.add_argument('--latitude', type=float, help='Set Faraday radio default latitude. Format \"ddmm.mmmm\"')
+parser.add_argument('--longitude', type=float, help='Set Faraday radio default longitude. Format \"dddmm.mmmm\"')
+parser.add_argument('--latitudedir', help='Set Faraday radio default latitude direction (N/S)')
+parser.add_argument('--longitudedir', help='Set Faraday radio default longitude direction (E/W)')
```
This allows other characters besides E/W. Filter?

kb1lqc committed:
Added configboot bitmask functionality

Added options to configure redledtx and configured bits and then functions
to implement a bitmask integer from these values.

kb1lqc commented:
One thing I noticed @kb1lqd 

![image](https://user-images.githubusercontent.com/331540/27773615-7b2ec3ea-5f32-11e7-82e8-94871fe0a365.png)

We treat the GPS bitmask and telemetry boot bitmask different. I am forced in python to create the integer value from bits for the configboot value but you accept discrete values in the configuration file for items like GPS boot and stuff. Should we be keeping these the same?

kb1lqd commented:
We should standardize this but not in this pr. More of a nice too have than
lacking functionality

On Jul 2, 2017 2:28 PM, "Bryce Salmi" <notifications@github.com> wrote:

> One thing I noticed @kb1lqd <https://github.com/kb1lqd>
>
> [image: image]
> <https://user-images.githubusercontent.com/331540/27773615-7b2ec3ea-5f32-11e7-82e8-94871fe0a365.png>
>
> We treat the GPS bitmask and telemetry boot bitmask different. I am forced
> in python to create the integer value from bits for the configboot value
> but you accept discrete values in the configuration file for items like GPS
> boot and stuff. Should we be keeping these the same?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FaradayRF/Faraday-Software/pull/219#issuecomment-312518065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUsyNx_xC-dAbemimcDFYo1NYuGfLCMks5sKAtngaJpZM4OKtOb>
> .
>

kb1lqc committed:
GPIO P3 bit operations implemented

Command line arguments now define the bits in GPIO p3 and a bitmask is
automatically generated from them.

kb1lqc committed:
Removed unnecessary print statements

Used for development and accidentally left in.

kb1lqc committed:
GPIO P4 Functionality Added

Command line arguments now set up GPIO P4 default values and create a
bitmask automatically.

kb1lqc committed:
Added GPPIO P5 Functionality

Added same functionality as GPIO P3 and P4 received to P5 this time. Also,
added telemetry interval checks for values greater than 0.

kb1lqc committed:
Cleaned up to pass pytest

Cleaned up PEP8 violations we check for so pytest would pass.

kb1lqc commented:
@kb1lqd I implemented the GPIO functionality as well as boot bit mask functionality. It's a tad brute forced in some regards so I'll make an issue to specify that we should clean this up and add functions to pragmatically handle the GPIO bitmask creation. Would be a good beginner task to keep open that doesn't affect functionality.

kb1lqd reviewed (approved):
Besides a few potential updates to clarify/improve in the future the code works as expected.

kb1lqc committed:
Fixed error in simpleconfig callsign

Callsign formatting was wrong for printing of callsign read from FLASH
memory. Corrected error.

kb1lqc committed:
Added bitmask decoding to simpleconfig

Decode the integers returned for bitmasks and print them out in human
readable format. There is a bug with setting redledtx that cuases the unit
to reset as if configboot was set...

kb1lqc committed:
Fixed redledtx and configured bit bug

Mixed up the bitmask order and therefore everytime you set --redledtxoff
you actually reconfigured the radio.

kb1lqc committed:
Cleaned up logging output for simpleconfig

Removed integer prints for bitmasks

kb1lqc committed:
Commented on bitmask code

Added comments to bitmask code for the read statement in simpleconfig.

kb1lqc committed:
Cleaned up comments

Added relevant headers to files for file info and licenses. Also added
helpful comments and docstrings.

kb1lqc committed:
Fixed error handling of simpleconfig

Now the program gracefully handles situations where
faraday-deviceconfiguration is not running.

kb1lqc committed:
Fixed Pytest warning

Removed blank tab on blank line

kb1lqc committed:
Updated error handling of incorrect proxy configs

Added valueerror exception to handle incorrectly configured proxy settings
in device configuration.
